File: app/src/main/java/com/example/groxweatherapp/grox/WeatherRequestCommand.java
package com.example.groxweatherapp.grox;

import com.example.groxweatherapp.api.WeatherApiClient;
import com.example.groxweatherapp.api.WeatherApiClient.RequestMode;
import com.groupon.grox.Action;
import com.groupon.grox.commands.rxjava1.Command;
import rx.Observable;
import rx.schedulers.Schedulers;

public class WeatherRequestCommand implements Command {

    private final int requestMode;

    public WeatherRequestCommand(@RequestMode int requestMode) {
        this.requestMode = requestMode;
    }

    @Override
    public Observable<Action> actions() {
        return new WeatherApiClient()
            .getWeather(requestMode)
            .subscribeOn(Schedulers.io())
            .map(weatherResponse -> new WeatherSuccessAction(weatherResponse, requestMode))
            .cast(Action.class)
            .onErrorReturn(throwable -> new WeatherErrorAction(throwable, requestMode))
            .startWith(new WeatherRefreshAction());
    }
}

File: app/src/main/java/com/example/groxweatherapp/grox/WeatherRequestCommand.kt
package com.example.groxweatherapp.grox

import com.example.groxweatherapp.api.WeatherApiClient
import com.example.groxweatherapp.api.WeatherApiClient.Companion.ForecastMode
import com.example.groxweatherapp.model.WeatherModel
import com.groupon.grox.Action
import com.groupon.grox.commands.rxjava1.Command
import rx.Observable
import rx.schedulers.Schedulers

class WeatherRequestCommand(@ForecastMode private var forecastMode: Int) : Command {

    override fun actions(): Observable<Action<WeatherModel>> {
        return WeatherApiClient()
                .getWeather(forecastMode)
                .subscribeOn(Schedulers.io())
                .map { weatherResponse -> WeatherSuccessAction(weatherResponse, forecastMode) as Action<WeatherModel>}
                .onErrorReturn { WeatherErrorAction(forecastMode) }
                .startWith(WeatherRefreshAction())
    }
}
